core/settings.ml

```diff
@@ -1,29 +1,139 @@
+   commands are ordered according to their discrete time in ascending
+   order.
+
+  *)
+
+module Printexc = Utility.Printexc
+
+type privilege = [`User | `System]
+
+(* Error handling. *)
+exception Insufficient_privileges of string
+exception Conversion_failure of exn * string * string
+exception Missing_argument of string
+exception Unknown_option of string
+exception Unknown_setting of string
+exception Invalid_keyvalue_pair of int * string
+exception Cannot_set_readonly of string
+
+let is_known_exn : exn -> bool = function
+  | Insufficient_privileges _
+  | Conversion_failure _
+  | Missing_argument _
+  | Unknown_option _
+  | Unknown_setting _
+  | Cannot_set_readonly _
+  | Invalid_keyvalue_pair _ -> true
+  | _ -> false
+
+let msg_of_exn : exn -> string option = function
+  | Missing_argument option_name ->
+     Some (Printf.sprintf "Missing argument for option '%s'" option_name)
+  | Conversion_failure (exn, setting_name, value) ->
+     Some (Printf.sprintf "Setting '%s': cannot parse value '%s' (%s)" setting_name value (Printexc.to_string exn))
+  | Insufficient_privileges setting_name ->
+     Some (Printf.sprintf "Cannot change system setting '%s'" setting_name)
+  | Unknown_option option_name ->
+     let option_name =
+       (if String.length option_name = 1
+        then "-" else "--") ^ option_name
+     in
+     Some (Printf.sprintf "Unknown option '%s'" option_name)
+  | Unknown_setting setting_name ->
+     Some (Printf.sprintf "Unknown setting '%s'" setting_name)
+  | Invalid_keyvalue_pair (_, entry) ->
+     Some (Printf.sprintf "Cannot parse '%s' as a key-value pair" entry)
+  | Cannot_set_readonly name ->
+     let kind =
+       if Char.equal '-' name.[0]
+       then "Option" else "Setting"
+     in
+     Some (Printf.sprintf "%s '%s' is read-only" kind name)
+  | _ -> None
+
+let classify : exn -> string = function
+  | Unknown_setting _
+  | Insufficient_privileges _ -> "warning"
+  | exn when is_known_exn exn -> "error"
+  | _ -> raise (Invalid_argument "classify")
+
+let print_exn : ?always_error:bool -> exn -> unit
+  = fun ?(always_error=false) exn ->
+  if is_known_exn exn
+  then ((if always_error
+         then Printf.fprintf stderr "error: %s\n" (Utility.val_of (msg_of_exn exn))
+         else Printf.fprintf stderr "%s: %s\n" (classify exn) (Utility.val_of (msg_of_exn exn)));
+        flush stderr)
+  else raise (Invalid_argument "print_exn")
+
+let default_handler : exn -> Printexc.raw_backtrace -> unit
+  = fun exn stacktrace ->
+  let open Utility in
+  if is_known_exn exn
+  then print_exn ~always_error:true exn
+  else let () = Printf.fprintf stderr "Fatal error: exception %s\n" (Printexc.to_string exn) in
+       let () = if Printexc.print_backtraces
+                then Printexc.print_raw_backtrace stderr stacktrace
+       in
+       flush stderr
+
+let _ = Printexc.set_uncaught_exception_handler default_handler
+
+let no_conv : string -> 'a
+  = fun _ -> failwith "No converter"
```

Comment:
I will see whether I can move the global exception handling stuff into `Errors` too....

Reply:
OK. So I looked into. It is rather difficult as it introduces a cyclic dependency between errors and settings, because `SourceCode` is included in `Errors`.

Reply:
So I will rather keep it as is for now, then maybe we can refactor it in the future. Also note, the `No_converter` error will be a graceful, it will be part of a parse failure message.

Reply:
OK, that's fine. I have run into this cyclic dependency issue a few times, so it's likely worth doing something about it, but that's for another issue. Thanks for checking.